### DISCUSSION TITLE
Finding memory leaks in solara with Memray and objgraph

### DISCUSSION THREAD

**Discussion by maartenbreddels** [widgetti/solara#172] [Show and tell]
We noticed some memory leaks in solara, but the technique I've used to find them can be very useful in general for solara applications (or any Python program).

The first step is to isolate the issue. Let's first render a component, without any server:
```python
import gc

import solara


@solara.component
def Page():
    for i in range(1000):
        solara.Button()


if __name__ == "__main__":
    box, rc = solara.render(Page(), handle_error=False)
    rc.close()

    del box
    del rc
    gc.collect()
```

We remove the reference to rc (the rendering context) and the container (`box)`, and run the garbage collector. I didn't expect that t manually running the garbage collector was needed, but we do need it for memray 🤷, which probably means I don't fully understand the lifecycle of memray and the Python process it runs.

Now we run it under [memray](https://github.com/bloomberg/memray):
```
$ PYTHONMALLOC=malloc python -m memray run -f -o render1k.bin --trace-python-allocators issues/memleak.py
riting profile results into render1k.bin
[memray] Successfully generated profile results.

You can now generate reports from the stored allocation records.
Some example commands to generate reports:

/Users/maartenbreddels/miniconda3/envs/dev/bin/python -m memray flamegraph render1k.bin
# we execute the suggested command to generate the flamegraph, but add the --leaks argument 
$ memray flamegraph --leaks render1k.bin -f
```

See https://bloomberg.github.io/memray/flamegraph.html for the `--leaks` argument.

This gives us:
<img width="1288" alt="image" src="https://github.com/widgetti/solara/assets/1765949/b5e7d6d3-580c-4f69-b893-a29bdfbe36a0">

Where we see on the right that indeed, solara is not releasing some memory, zooming in we see this is using 11Mb of memory:
<img width="1278" alt="image" src="https://github.com/widgetti/solara/assets/1765949/b63129e1-1153-44e0-bd5a-b31e837e5531">

My guess here is that the `rc` (render context), has an extra reference to it. To check this, we use [objgraph](https://objgraph.readthedocs.io/en/stable/)

```python
import gc

import solara


@solara.component
def Page():
    for i in range(1000):
        solara.Button()


if __name__ == "__main__":
    import objgraph

    import solara

    box, rc = solara.render(Page(), handle_error=False)
    rc.close()
    objgraph.show_backrefs([rc], filename="solara-mem-leak.png")

    del rc
    del box
    gc.collect()
```

We run the program (with or without memray, doesn't matter), and inspect `solara-mem-leak.png`:

![image](https://github.com/widgetti/solara/assets/1765949/065a46c2-4ea3-4665-8730-3e11f301fafd)

On the right, we see many internal references, but on the left we see the `rc` reference from our script *and* a reference from a thread-local object!.

Fixing this in Reacton (the underlying library of solara) in  https://github.com/widgetti/reacton/commit/8ebc3a43725e9768827605339730e4df8124700a we see no more significant memory usage due to solara:
<img width="1345" alt="image" src="https://github.com/widgetti/solara/assets/1765949/cabb33bd-7f9e-4b5e-ad2c-8d8ae64bcf9c">

Hope this helps people debug memory leak issues! Ideas or improvements on this workflow are welcome 🙏

**Comment by pythonwood** [comment]
thanks, learn much. I test in solara==1.57.3 .  it is so simple but jsut not in tree node  "import solara" now.

<img width="1211" height="656" alt="image" src="https://github.com/user-attachments/assets/6fc50f62-1eed-48fe-a001-4e41363d8f50" />

<img width="581" height="417" alt="image" src="https://github.com/user-attachments/assets/79d698dc-142a-4ec7-9cdb-5f379868f85e" />
